src/persist/src/operators/stream.rs

```diff
@@ -359,10 +361,14 @@ where
                 input_frontier.clone_from(&new_input_frontier);
                 // We need to downgrade when the input frontier is empty. This basically releases
                 // all the capabilities so that downstream operators and eventually the worker can
-                // shut down.
+                // shut down. We also need to clear all pending futures to make sure we never
+                // attempt to downgrade any more capabilities.
                 if input_frontier.is_empty() {
                     cap_set.downgrade(input_frontier.iter());
+                    // pending_futures.clear();
```

Comment:
hang on a sec. I'm confused because if we drop the capabilities we need to clear `pending_future`s (or do a check when we go to `seal`)

Reply:
Ok phew. good. tests break as I had expected. I believe this commit makes tests simpler but forces the operator to be more complicated and I'm not sure its worth it. Let me chew on this a bit more.

Reply:
Aljoscha and I talked over slack. This commit was meant to illustrate that the tests still break if you comment out the fix. With both of our changes combined I think everything works (Aljoscha's changes make it so that its easier to test what happens when you close an operator, and the operator is still able to clear its futures when it wants to exit immediately)